src/broker/Manager.cc

```diff
@@ -92,6 +92,142 @@ void print_escaped(std::string& buf, std::string_view str) {
 }
 
+// Track metrics for a given peering's send buffer.
+class PeerBufferState {
+public:
+    struct Stats {
+        Stats(const broker::endpoint_id& peer) {
+            std::string peer_s;
+            broker::convert(peer, peer_s);
+            peer_id = zeek::make_intrusive<zeek::StringVal>(peer_s);
+        }
+
+        // The rendered peer ID. Storing this here helps reuse.
+        zeek::StringValPtr peer_id;
+
+        // Number of messages queued locally in the send buffer.
+        uint32_t queued = 0;
+
+        // Maximum number queued in the last Broker::buffer_stats_interval.
+        // This improces visibility into message bursts since instantaneous
+        // queueing (captured above) can be short-lived.
+        uint32_t max_queued_recently = 0;
+
+        // Number of times the buffer overflowed at send time.  For the
+        // "disconnect" overflow policy (via Broker::peer_overflow_policy), this
+        // count will at most be 1 since Broker will remove the peering upon
+        // overflow. The existing Zeek-level metric for tracking disconnects
+        // (see frameworks/broker/broker-backpressure.zeek) covers this one more
+        // permanently. For the "drop_newest" and "drop_oldest" policies it
+        // equals a count of the number of messages lost, since the peering
+        // continues.
+        uint64_t overflows = 0;
+
+        // When we last started a stats-tracking interval for this peering.
+        double last_interval = 0;
+    };
+
+    // For per-peering tracking, map endpoint IDs to the above state.
+    using EndpointMetricMap = std::unordered_map<broker::endpoint_id, Stats>;
+
+    PeerBufferState(size_t a_buffer_size, double a_stats_interval)
+        : buffer_size(a_buffer_size), stats_interval(a_stats_interval) {
+        stats_table =
+            zeek::make_intrusive<zeek::TableVal>(zeek::id::find_type<zeek::TableType>("BrokerPeeringStatsTable"));
+        stats_record_type = zeek::id::find_type<zeek::RecordType>("BrokerPeeringStats");
+    }
+
+    // Update the peering's stats. This runs in Broker's execution context.
+    // Broker does not expose send-buffer/queue state explicitly, so track
+    // arrivals (a push, push_or_pull == true) and departures (pull,
+    // push_or_pull == false) as they happen.
+    void Observe(const broker::endpoint_id& peer, bool push_or_pull) {
+        std::lock_guard<std::mutex> lock(mutex);
+        auto it = stats_map.find(peer);
+
+        if ( it == stats_map.end() ) {
+            Stats stats{peer};
+            stats_map.emplace(peer, stats);
+            stats_table->Assign(stats.peer_id, zeek::make_intrusive<zeek::RecordVal>(stats_record_type));
+            it = stats_map.find(peer);
+        }
+
+        auto& stats = it->second;
+        auto now = zeek::util::current_time();
+
+        if ( now - stats.last_interval > stats_interval ) {
+            stats.last_interval = now;
+            stats.max_queued_recently = 0;
```

Comment:
```suggestion
            stats.max_queued_recently = stats.queued;
```

Not sure that matters all that much, but given we may not update `max_queued_recently` below if we're in a overflow situation, might make sense.